### PR TITLE
Faster traverse/sequence using left fold.

### PR DESCRIPTION
**Pull request by jacereda**

### CODE REVIEW

**Comment by jacereda** [comment]
Also, it seems something like this runs out of memory with the previous implementation (at least on node):

    module Test.Main where
    import Debug.Trace
    import Data.Maybe
    import Data.Array
    import Data.Traversable

    main = do
      let a = map Just (range 1 100000)
      let x = sequence a
      trace "done"
      return unit

File: src/Data/Traversable.purs [hunk 1]
```diff
@@ -43,11 +43,19 @@ class (Functor t, Foldable t) <= Traversable t where
...
+
 instance traversableArray :: Traversable [] where
-  traverse _ []     = pure []
-  traverse f (x:xs) = (:) <$> (f x) <*> traverse f xs
-  sequence []     = pure []
-  sequence (x:xs) = (:) <$> x <*> sequence xs
+  traverse f xs = sequence (f <$> xs)
+  sequence = foldl step (pure [])
+    where step :: m [a] -> m a -> m [a]
+          step sofar x = apush <$> sofar <*> x
```
Comment (paf31) [inline src/Data/Traversable.purs]: Can we prove that `apush` is safe for all `Applicative`s?

Comment (jacereda) [inline src/Data/Traversable.purs]: I don't have an answer for that... @garyb?

Comment (paf31) [inline src/Data/Traversable.purs]: I strongly suspect the answer is "no". Try the list applicative for example.

Comment (jacereda) [inline src/Data/Traversable.purs]: Right, I'll close it.

Comment (jacereda) [inline src/Data/Traversable.purs]: What about something like this? It isn't a speed daemon but at least the above example doesn't crash...

    sequence = foldr step (pure [])
      where step x sofar = (:) <$> x <*> sofar

Comment (jacereda) [inline src/Data/Traversable.purs]: Damn, I think I messed it... it should probably be a left fold using snoc.

**Pull request closed without merging**